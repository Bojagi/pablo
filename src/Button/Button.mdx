---
name: Button
route: /button
menu: Components
---

import { Props, Playground } from 'docz'
import { Button } from './Button.tsx'
import { Box } from '../Box/Box.tsx';
import { css, keyframes } from '@emotion/react';
import { GitPullRequest } from 'react-feather';

# Button

Interactive element that can either trigger an action or can be used as a link.

## Examples

### Different sizes

Buttons can be either `small`, `medium` or `large`. Default size is `medium`.

<Playground>
  <Box mx={-4}>  
    <Button size="small" mx={4}>
      Small
    </Button>
    <Button mx={4}>
      Medium
    </Button>
    <Button size="large" mx={4}>
      Large
    </Button>
  </Box>
</Playground>

### Colors

Colors can be used to give a button a certain emphasis. By default a button uses the `brand` color.
It can also be set to `plain` for actions that are not as important. If you want show that a button
triggers a descructive action (like deleting something) you can use `negative`. 
You can also use the color `positive` for construcive or positive actions.

<Playground>
  <Box mx={-4}>  
    <Button mx={4}>
      Brand button
    </Button>
    <Button color="plain" mx={4}>
      Plain button
    </Button>
    <Button color="positive" mx={4}>
      Positive button
    </Button>
    <Button color="negative" mx={4}>
      Negative button
    </Button>
  </Box>
</Playground>

### Variants

In addition to color you can use different variants. Variants are an even better way to structure your
buttons and give them different levels of importance. By default a button has the variant `primary`.
The variant `secondary` and `text` can be used if you want to give a button less importance, 
where `secondary` is more important than `text`.

<Playground>
  <Box mx={-4}>  
    <Button mx={4}>
      Accept
    </Button>
    <Button variant="secondary" mx={4}>
      Cancel
    </Button>
    <Button variant="text" mx={4}>
      read more
    </Button>
  </Box>
</Playground>

### Inverted variants

Similar to variants, you can use inverted variants. These are useful if you want to show the variants
on brand colored or dark background. For this you need to add the `Inverted` suffix on the variant prop,
making it camel case (e.g. `primaryInverted`).

<Playground>
  <Box borderRadius={4} py={6} px={4}  bgColor="brand.main">  
    <Button variant="primaryInverted" mx={4}>
      Accept
    </Button>
    <Button variant="secondaryInverted" mx={4}>
      Cancel
    </Button>
    <Button variant="textInverted" mx={4}>
      read more
    </Button>
  </Box>
</Playground>

### Icons

If you want to make your button more recognizable or give it even more emphasis you can use icons.
Icons can be either added to the start or end of a button. You can use any icon library you want
as Pablo is agnostic regarding libraries. In the examples below we use icons from `react-feather`.
Colors are applied based on the `variant` and `color` prop.

<Playground>
  <Box display="flex" alignItems="center" mx={-4}>
    <Button mx={4} endIcon={<GitPullRequest size={16} />}>
      Accept
    </Button>
    <Button mx={4} startIcon={<GitPullRequest size={16} />}>
      Cancel
    </Button>
    <Button mx={4} variant="secondary" endIcon={<GitPullRequest size={16} />}>
      Accept
    </Button>
    <Button mx={4} variant="secondary" startIcon={<GitPullRequest size={16} />}>
      Cancel
    </Button>
  </Box>
</Playground>

### Using buttons as link

You can use the `as` property to change the component of the button. 
In the example below we use simple link, but you can use the `Link` component 
of `react-router` or `gatsby` as well.

<Playground>
  <Button as="a" href="https://bojagi.io" target="_blank">
    Link button
  </Button>
</Playground>

### Using full width buttons
If you want your button to span over the full width of it's parent, you can use the `fullWidth` boolean prop.

<Playground>
  <Button fullWidth endIcon={<GitPullRequest size={16} />}>
    I am a full width button
  </Button>
</Playground>

### Customize a button

You can customize every bit of a button with the `customStyles` prop using the `css` function of `styled-component`.

You can customize following things: 

* `icon` (general icon styles)
* `startIcon`
* `endIcon`
* `primary`
* `secondary`
* `text`

<Playground>
  <Box display="flex" justifyContent="center">
    <Button startIcon={<GitPullRequest size={16} />} customStyles={{
      primary: css`
        transform: rotate(4deg);
        background-color: tomato;
        &:hover:enabled {
          background-color: red;
        }
      `,
      icon: css`
        color: blue;
      `,
      startIcon: css`
        margin-right: 100px;
      `,
    }}>
      I am fancy
    </Button>
  </Box>
</Playground>



## Import

```js
import { Button } from '@bojagi/pablo/Button';
```

## Props
<Props of={Button} />
