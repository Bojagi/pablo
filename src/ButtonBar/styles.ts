import { getSpacing } from '../styleHelpers/getSpacing';
import { Style } from '../theme/types';

export interface ButtonBarStyles {
  gap: Style;
}

export const buttonBarStyles: ButtonBarStyles = {
  gap: getSpacing(2),
};
