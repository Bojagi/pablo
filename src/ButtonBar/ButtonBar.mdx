---
name: ButtonBar
route: /buttonbar
menu: Components
---

import { Props, Playground } from 'docz'
import { ButtonBar } from './ButtonBar.tsx'
import { Button } from '../Button/Button.tsx'
import { Card } from '../Card/Card.tsx';
import { Title, Paragraph } from '../Typography/index.ts';
import { Box } from '../Box/Box.tsx';

# ButtonBar

A layouting component to put multiple buttons next to each other aligned to the right. 
It's handy for dialogs and confirmations.

## Example

<Playground>
  <Card width={500}>
    <Title>Do you want to proceed?</Title>
    <Paragraph>Please select one of the buttons below!</Paragraph>
    <ButtonBar>
      <Button size="medium" variant="text" color="plain">
        Cancel
      </Button>
      <Button size="medium">Accept</Button>
    </ButtonBar>
  </Card>
</Playground>

## Import

```js
import { ButtonBar } from '@bojagi/pablo/ButtonBar';
```

## Props
<Props of={ButtonBar} />
