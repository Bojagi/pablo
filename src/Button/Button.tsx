import React, { forwardRef } from 'react';
import { buttonBaseStyles, ButtonBaseProps, ButtonSize } from '../ButtonBase';
import styled from '@emotion/styled';
import { getComponentStyle } from '../styleHelpers/getComponentStyle';
import { BaseProps, CssFunctionReturn } from '../types';
import { ButtonTypography } from '../Typography';
import { useCustomStyles } from '../utils/useCustomStyles';
import { ButtonStyleProperties } from './styles';
import { interpolateCssProp } from '../utils/interpolateCssProp';

export type ButtonVariant =
  | 'primary'
  | 'primaryInverted'
  | 'secondary'
  | 'secondaryInverted'
  | 'text'
  | 'textInverted';
export type ButtonColor = 'brand' | 'plain' | 'negative' | 'positive';

export interface InnerButtonProps extends ButtonBaseProps, BaseProps<ButtonStyleProperties> {
  children: React.ReactNode;
  variant?: ButtonVariant;
  fullWidth?: boolean;
  innerRef?: React.Ref<HTMLButtonElement>;
  startIcon?: React.ReactNode;
  endIcon?: React.ReactNode;
  color?: ButtonColor;
  disabled?: boolean;
  className?: string;
  cssStyles?: CssFunctionReturn;
  ref?: React.Ref<any>;
}

export type ButtonProps<C extends React.ElementType> = InnerButtonProps &
  React.ComponentPropsWithRef<C>;

const getColorStyles = (props, suffix?: string) => {
  const basePath = 'button.{color}.{variant}';
  const suffixedPath = [basePath, suffix].filter((x) => x).join('.');

  return {
    color: getComponentStyle(`${suffixedPath}.color`)(props),
    borderColor: getComponentStyle(`${suffixedPath}.borderColor`)(props),
    background: getComponentStyle(`${suffixedPath}.backgroundColor`)(props),
  };
};

const InnerButton = styled('button')<InnerButtonProps>([
  buttonBaseStyles,
  (props) => ({
    justifyContent: 'center',
    width: props.fullWidth ? '100%' : 'inherit',
    ...getColorStyles(props),
    '&:focus': {
      boxShadow: `0 0 0 ${getComponentStyle('button.base.focus.outlineSize')(
        props
      )} ${getComponentStyle('button.{color}.outlineColor')(props)}`,
    },
    '&:hover:enabled': getColorStyles(props, 'hover'),
    '&:active:enabled': getColorStyles(props, 'active'),
  }),
]);

interface IconBoxProps {
  size: ButtonSize;
  marginSide: 'left' | 'right';
  css?: CssFunctionReturn;
  children?: React.ReactNode;
}

const IconBox = styled.div((props: IconBoxProps) => [
  {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    [`margin-${props.marginSide}`]: getComponentStyle('button.base.icon.gap')(props),

    '&, & > *': {
      width: getComponentStyle('button.base.icon.size.{size}')(props),
      height: getComponentStyle('button.base.icon.size.{size}')(props),
    },
  },
  interpolateCssProp,
]);

const knownVariants = [
  'primary',
  'primaryInverted',
  'secondary',
  'secondaryInverted',
  'text',
  'textInverted',
];

export const Button = forwardRef<HTMLButtonElement, ButtonProps<any>>(
  (
    {
      color = 'brand',
      variant = 'primary',
      size = 'medium',
      startIcon,
      endIcon,
      children,
      className,
      disabled,
      customStyles,
      fullWidth = false,
      ...props
    },
    ref
  ) => {
    const isKnownVariant = knownVariants.indexOf(variant) > -1;
    const getCustomStyles = useCustomStyles('button.styles', customStyles);

    return (
      <InnerButton
        data-testid="pbl-button"
        variant={isKnownVariant ? variant : 'primary'}
        {...props}
        ref={ref}
        size={size}
        className={className}
        disabled={disabled}
        color={color}
        fullWidth={fullWidth}
        css={getCustomStyles(variant)}
      >
        {startIcon && (
          <IconBox
            css={(props) => [getCustomStyles('icon', props), getCustomStyles('startIcon', props)]}
            marginSide="right"
            size={size}
            data-testid="pbl-button-icon"
          >
            {startIcon}
          </IconBox>
        )}
        <ButtonTypography>{children}</ButtonTypography>
        {endIcon && (
          <IconBox
            css={(props) => [getCustomStyles('icon', props), getCustomStyles('endIcon', props)]}
            marginSide="left"
            size={size}
            data-testid="pbl-button-icon"
          >
            {endIcon}
          </IconBox>
        )}
      </InnerButton>
    );
  }
);
