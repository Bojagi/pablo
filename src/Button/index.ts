import './styles';

export * from './Button';
