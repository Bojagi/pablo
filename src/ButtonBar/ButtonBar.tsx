import React, { forwardRef, cloneElement } from 'react';
import styled from '@emotion/styled';
import { BoxProps, LayoutBoxProps } from '../Box';
import { useComponentStyle } from '../theme/useComponentStyle';
import { guaranteeArray } from '../utils/guaranteeArray';
import { getComponentStyle } from '../styleHelpers';
import { getCustomStyles } from '../utils/useCustomStyles';
import { ButtonBarStyleProperties } from './styles';
import { BaseProps } from '../types';
import { baseStyle } from '../shared/baseStyle';

export interface ButtonBarProps extends BaseProps<ButtonBarStyleProperties>, LayoutBoxProps {
  children: React.ReactElement<BoxProps> | React.ReactElement<BoxProps>[];
}

const ButtonBarBox = styled.div`
  ${baseStyle}
  display: flex;
  justify-content: flex-end;
  margin: 0 -${getComponentStyle('buttonBar.gap')};
  ${getCustomStyles('buttonBar.styles', 'root')}
`;

export const ButtonBar = forwardRef<HTMLDivElement, ButtonBarProps>(
  ({ children, ...props }: ButtonBarProps, ref) => {
    const gap = useComponentStyle('buttonBar.gap');
    return (
      <ButtonBarBox ref={ref} data-testid="pbl-buttonbar" {...props}>
        {guaranteeArray(children).map((child, i) =>
          cloneElement(child, {
            key: i,
            mx: gap,
          })
        )}
      </ButtonBarBox>
    );
  }
);
