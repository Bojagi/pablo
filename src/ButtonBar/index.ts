import './styles';

export * from './ButtonBar';
