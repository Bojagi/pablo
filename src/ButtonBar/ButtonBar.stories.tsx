import React from 'react';
import { css } from 'styled-components';
import { Button } from '../Button';
import { Card } from '../Card';
import { Paragraph, Title } from '../Typography';
import { ButtonBar } from './ButtonBar';

export default {
  title: 'ButtonBar',
};

export const WithTwoButtons = () => (
  <Card width={500}>
    <Title>Do you want to proceed?</Title>
    <Paragraph>Please select one of the buttons below!</Paragraph>
    <ButtonBar>
      <Button size="medium" variant="text" color="plain">
        Cancel
      </Button>
      <Button size="medium">Accept</Button>
    </ButtonBar>
  </Card>
);

export const WithOneButtons = () => (
  <Card width={500}>
    <Title>Move along</Title>
    <Paragraph>You don't have an option!</Paragraph>
    <ButtonBar>
      <Button size="medium">Ok</Button>
    </ButtonBar>
  </Card>
);

export const WithCustomStyles = () => (
  <Card width={500}>
    <Title>Do you want to proceed?</Title>
    <Paragraph>Please select one of the buttons below!</Paragraph>
    <ButtonBar
      customStyles={{
        root: css`
          padding: 7px;
          background-color: red;
        `,
      }}
    >
      <Button size="medium" variant="text" color="plain">
        Cancel
      </Button>
      <Button size="medium">Accept</Button>
    </ButtonBar>
  </Card>
);
