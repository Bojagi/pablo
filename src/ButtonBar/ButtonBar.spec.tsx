import { render } from '@testing-library/react';
import React from 'react';
import { Button } from '../Button/Button';
import { PabloThemeProvider } from '../theme';
import { ButtonBar } from './ButtonBar';

test('Render button bar', () => {
  const { container, getByTestId, getAllByTestId } = renderComponent(
    <ButtonBar>
      <Button>Button 1</Button>
      <Button>Button 2</Button>
    </ButtonBar>
  );
  expect(container).toMatchSnapshot();
  expect(getByTestId('pbl-buttonbar')).toHaveStyleRule('gap', '0.25rem 0.25rem');
  const allButtons = getAllByTestId('pbl-button');
  expect(allButtons).toBeArrayOfSize(2);
});

test('Render button bar with one item', () => {
  const { container } = renderComponent(
    <ButtonBar>
      <Button>Button 1</Button>
    </ButtonBar>
  );
  expect(container).toMatchSnapshot();
});

function renderComponent(elem) {
  return render(<PabloThemeProvider>{elem}</PabloThemeProvider>);
}
