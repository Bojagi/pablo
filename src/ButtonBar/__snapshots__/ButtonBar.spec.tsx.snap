// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render button bar 1`] = `
.emotion-0 {
  box-sizing: border-box;
  gap: 0.25rem 0.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
}

.emotion-1 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 10px 14px;
  border: 1px solid transparent;
  background: transparent;
  border-radius: 0.25rem;
  -webkit-transition: background-color,0.3s,color 0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out,-webkit-transform,0.15s;
  transition: background-color,0.3s,color 0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out,transform,0.15s;
  outline: none;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: inherit;
  color: var(--pbl-theme-colors-brand-contrastText);
  border-color: var(--pbl-theme-colors-brand-main);
  background: var(--pbl-theme-colors-brand-main);
}

.emotion-1:enabled {
  cursor: pointer;
}

.emotion-1:disabled {
  opacity: 0.4;
  cursor: normal;
}

.emotion-1:focus {
  box-shadow: 0 0 0 0.25rem var(--pbl-theme-colors-brand-lightest);
}

.emotion-1:hover:enabled {
  color: var(--pbl-theme-colors-brand-contrastText);
  border-color: var(--pbl-theme-colors-brand-dark);
  background: var(--pbl-theme-colors-brand-dark);
}

.emotion-1:active:enabled {
  color: var(--pbl-theme-colors-brand-contrastText);
  border-color: var(--pbl-theme-colors-brand-darkest);
  background: var(--pbl-theme-colors-brand-darkest);
}

.emotion-2 {
  box-sizing: border-box;
  font-family: var(--pbl-theme-typography-base-fontFamily);
  font-style: normal;
  font-weight: var(--pbl-theme-typography-base-fontWeight);
  margin: 0;
  font-size: var(--pbl-theme-typography-button-fontSize);
  line-height: var(--pbl-theme-typography-button-lineHeight);
  margin-bottom: var(--pbl-theme-typography-button-marginBottom);
}

<div>
  <div
    class="emotion-0"
    data-testid="pbl-buttonbar"
  >
    <button
      class="emotion-1"
      color="brand"
      data-testid="pbl-button"
    >
      <span
        class="emotion-2"
      >
        Button 1
      </span>
    </button>
    <button
      class="emotion-1"
      color="brand"
      data-testid="pbl-button"
    >
      <span
        class="emotion-2"
      >
        Button 2
      </span>
    </button>
  </div>
</div>
`;

exports[`Render button bar with one item 1`] = `
.emotion-0 {
  box-sizing: border-box;
  gap: 0.25rem 0.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
}

.emotion-1 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 10px 14px;
  border: 1px solid transparent;
  background: transparent;
  border-radius: 0.25rem;
  -webkit-transition: background-color,0.3s,color 0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out,-webkit-transform,0.15s;
  transition: background-color,0.3s,color 0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out,transform,0.15s;
  outline: none;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: inherit;
  color: var(--pbl-theme-colors-brand-contrastText);
  border-color: var(--pbl-theme-colors-brand-main);
  background: var(--pbl-theme-colors-brand-main);
}

.emotion-1:enabled {
  cursor: pointer;
}

.emotion-1:disabled {
  opacity: 0.4;
  cursor: normal;
}

.emotion-1:focus {
  box-shadow: 0 0 0 0.25rem var(--pbl-theme-colors-brand-lightest);
}

.emotion-1:hover:enabled {
  color: var(--pbl-theme-colors-brand-contrastText);
  border-color: var(--pbl-theme-colors-brand-dark);
  background: var(--pbl-theme-colors-brand-dark);
}

.emotion-1:active:enabled {
  color: var(--pbl-theme-colors-brand-contrastText);
  border-color: var(--pbl-theme-colors-brand-darkest);
  background: var(--pbl-theme-colors-brand-darkest);
}

.emotion-2 {
  box-sizing: border-box;
  font-family: var(--pbl-theme-typography-base-fontFamily);
  font-style: normal;
  font-weight: var(--pbl-theme-typography-base-fontWeight);
  margin: 0;
  font-size: var(--pbl-theme-typography-button-fontSize);
  line-height: var(--pbl-theme-typography-button-lineHeight);
  margin-bottom: var(--pbl-theme-typography-button-marginBottom);
}

<div>
  <div
    class="emotion-0"
    data-testid="pbl-buttonbar"
  >
    <button
      class="emotion-1"
      color="brand"
      data-testid="pbl-button"
    >
      <span
        class="emotion-2"
      >
        Button 1
      </span>
    </button>
  </div>
</div>
`;
